// devextreme/cjs/__internal/scheduler/r1/components/base/group_panel_props.js
/**
 * DevExtreme (cjs/__internal/scheduler/r1/components/base/group_panel_props.js)
 * Version: 23.2.6
 * Build date: Wed May 01 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.GroupPanelRowDefaultProps = exports.GroupPanelCellDefaultProps = exports.GroupPanelBaseDefaultProps = void 0;
const GroupPanelBaseDefaultProps = {
    groupPanelData: {
        groupPanelItems: [],
        baseColSpan: 1
    },
    groupByDate: false,
    styles: {}
};
exports.GroupPanelBaseDefaultProps = GroupPanelBaseDefaultProps;
const GroupPanelCellDefaultProps = {
    id: 0,
    text: "",
    data: {
        id: 0
    },
    className: ""
};
exports.GroupPanelCellDefaultProps = GroupPanelCellDefaultProps;
const GroupPanelRowDefaultProps = {
    groupItems: [],
    className: ""
};
exports.GroupPanelRowDefaultProps = GroupPanelRowDefaultProps;
