// devextreme/cjs/__internal/scheduler/r1/const.js
/**
 * DevExtreme (cjs/__internal/scheduler/r1/const.js)
 * Version: 23.2.6
 * Build date: Wed May 01 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.VIEW_TYPES = exports.VERTICAL_GROUP_ORIENTATION = exports.TIMELINE_VIEWS = exports.HORIZONTAL_GROUP_ORIENTATION = void 0;
const VERTICAL_GROUP_ORIENTATION = "vertical";
exports.VERTICAL_GROUP_ORIENTATION = "vertical";
const HORIZONTAL_GROUP_ORIENTATION = "horizontal";
exports.HORIZONTAL_GROUP_ORIENTATION = "horizontal";
const TIMELINE_VIEWS = {
    timelineDay: true,
    timelineWeek: true,
    timelineWorkWeek: true,
    timelineMonth: true
};
exports.TIMELINE_VIEWS = TIMELINE_VIEWS;
const VIEW_TYPES = ["day", "week", "workWeek", "month", "timelineDay", "timelineWeek", "timelineWorkWeek", "timelineMonth", "agenda"];
exports.VIEW_TYPES = VIEW_TYPES;
